pkg/services/ngalert/state/state_tracker.go

```diff
@@ -24,77 +31,132 @@ type cache struct {
 		}
-		st.stateCache.update(currentState)
 	}
+	st.Log.Debug("returning changed states to scheduler", "count", len(changedStates))
 	return changedStates
 }
 
-func (st *StateTracker) getNextState(uid string, result eval.Result) eval.State {
-	currentState := st.stateCache.getOrCreate(uid, result)
-	if currentState.State == result.State {
-		return currentState.State
-	}
-
+//Set the current state based on evaluation results
+//return the state and a bool indicating whether a state transition occurred
+func (st *StateTracker) setNextState(uid string, result eval.Result) (AlertState, bool) {
+	currentState := st.getOrCreate(uid, result)
+	st.Log.Debug("setting alert state", "uid", uid)
 	switch {
 	case currentState.State == result.State:
-		return currentState.State
+		st.Log.Debug("no state transition", "cacheId", currentState.CacheId, "state", currentState.State.String())
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, false
 	case currentState.State == eval.Normal && result.State == eval.Alerting:
-		return eval.Alerting
+		st.Log.Debug("state transition from normal to alerting", "cacheId", currentState.CacheId)
+		currentState.State = eval.Alerting
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.StartsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt.Add(40 * time.Second))
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	case currentState.State == eval.Alerting && result.State == eval.Normal:
-		return eval.Normal
+		st.Log.Debug("state transition from alerting to normal", "cacheId", currentState.CacheId)
+		currentState.State = eval.Normal
+		currentState.EvaluatedAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.EndsAt = strfmt.DateTime(result.EvaluatedAt)
+		currentState.Results = append(currentState.Results, result.State)
+		st.set(currentState)
+		return currentState, true
 	default:
-		return eval.Alerting
+		return currentState, false
+	}
+}
+
+func (st *StateTracker) cleanUp() {
+	ticker := time.NewTicker(time.Duration(60) * time.Minute)
+	st.Log.Debug("starting cleanup process", "intervalMinutes", 60)
+	wg := sync.WaitGroup{}
+	for {
+		select {
+		case <-ticker.C:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				st.trim()
+			}()
+			wg.Wait()
+		case <-st.quit:
+			st.Log.Debug("stopping cleanup process", "now", time.Now())
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+func (st *StateTracker) trim() {
+	st.Log.Info("trimming alert state cache", "now", time.Now())
+	st.stateCache.mu.Lock()
+	defer st.stateCache.mu.Unlock()
+	for _, v := range st.stateCache.cacheMap {
+		if len(v.Results) > 100 {
+			st.Log.Debug("trimming result set", "cacheId", v.CacheId, "count", len(v.Results)-100)
+			v.Results = v.Results[:len(v.Results)-100]
```

Comment:
This is trimming the latest added states, should this be `v.Results[100:]`?

Q1: if we are trimming only 100 every hour, isn't there a chance of this slice growing forever if we add more than 100 results per hour?

Q2: If my above reasoning about trimming from index 0 is right, would Go internally hold onto the trimmed piece of slice and should we recycle the slice (i.e. copy back items to index 0) instead?

Reply:
Agree on both points.